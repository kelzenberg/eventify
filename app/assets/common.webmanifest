{
    "icons": [
        { "src": "/assets/Favicon_192.png", "type": "image/png", "sizes": "192x192" },
        { "src": "/assets/Favicon_512.png", "type": "image/png", "sizes": "512x512" }
    ]
}