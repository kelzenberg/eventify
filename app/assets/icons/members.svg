<svg width="19" height="12" viewBox="0 0 19 12" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M12.7742 5.18183C14.1324 5.18183 15.2206 4.08546 15.2206 2.72728C15.2206 1.3691 14.1324 0.272736 12.7742 0.272736C11.416 0.272736 10.3196 1.3691 10.3196 2.72728C10.3196 4.08546 11.416 5.18183 12.7742 5.18183ZM6.22874 5.18183C7.58692 5.18183 8.6751 4.08546 8.6751 2.72728C8.6751 1.3691 7.58692 0.272736 6.22874 0.272736C4.87056 0.272736 3.77419 1.3691 3.77419 2.72728C3.77419 4.08546 4.87056 5.18183 6.22874 5.18183ZM6.22874 6.81819C4.32237 6.81819 0.501465 7.77546 0.501465 9.68183V11.7273H11.956V9.68183C11.956 7.77546 8.1351 6.81819 6.22874 6.81819ZM12.7742 6.81819C12.5369 6.81819 12.2669 6.83455 11.9806 6.8591C12.9296 7.54637 13.5924 8.47092 13.5924 9.68183V11.7273H18.5015V9.68183C18.5015 7.77546 14.6806 6.81819 12.7742 6.81819Z" fill="#B1B1B4"/>
</svg>
