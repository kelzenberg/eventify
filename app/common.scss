@use "sass:map";
@import "./bootstrap.custom.scss";

body {
    margin: 0px;
    font-family: 'Athiti', sans-serif;
}

.fancy-background {
    position: fixed;
    top: 0px;
    left: 0px;
    right: 0px;
    bottom: 0px;
    z-index: -5;
    overflow: hidden;
    background: linear-gradient(160deg, #F6CAA2, #F2B680);

    &::before {
        content: "";
        position: absolute;
        z-index: 0;
        top: 0px;
        left: 0px;
        width: 200vw;
        height: 30vh;
        transform: rotateZ(-20deg);
        transform-origin: 0% 100%;
        background: #f5ceaa;
        filter: blur(5px);
    }

    &::after {
        content: "";
        position: absolute;
        z-index: 0;
        bottom: 0px;
        right: 0px;
        width: 200vw;
        height: 30vh;
        transform: rotateZ(-20deg);
        transform-origin: 100% 0%;
        background: #F6CAA2;
        filter: blur(5px);
    }
}

.iconBox {
    display: inline-block;
    position: relative;
    border-radius: 13px;
    border: 1px solid map.get($theme-colors, "primary");
    width: 50px;
    height: 50px;
    background-repeat: no-repeat;
    background-position: 50% 50%;
    vertical-align: middle;

    img {
        width: 100%;
        height: 100%;
    }

    @each $color, $value in $theme-colors {
        &.iconBox-#{$color} {
            border-color: $value;
        }
    }

    &.iconBox-removable:hover::after {
        content: "";
        position: absolute;
        top: 0px;
        left: 0px;
        right: 0px;
        bottom: 0px;
        border-radius: 13px;
        background-color: rgba(255, 255, 255, 0.5);
        background-image: url("/assets/icons/remove.svg");
        background-repeat: no-repeat;
        background-position: 50% 50%;
        pointer-events: none;
    }
}

.hover-accent:hover {
    background-color: #ececec;
}