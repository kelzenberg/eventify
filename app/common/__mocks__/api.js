const api = jest.requireActual("../api");

const UnexpectedServerError = api.default.UnexpectedServerError;

const authenticate = jest.fn().mockResolvedValue({});
const getUserInfo = jest.fn().mockResolvedValue({});
const register = jest.fn().mockResolvedValue({});
const completeRegistration = jest.fn().mockResolvedValue({});
const getAllUserEvents = jest.fn().mockResolvedValue({});
const createEvent = jest.fn().mockResolvedValue({});
const getEvent = jest.fn().mockResolvedValue({});
const saveEvent = jest.fn().mockResolvedValue({});
const leaveEvent = jest.fn().mockResolvedValue({});
const inviteToEvent = jest.fn().mockResolvedValue({});
const bounceFromEvent = jest.fn().mockResolvedValue({});
const addExpenseSharingModule = jest.fn().mockResolvedValue({});
const addPaymentToExpenseSharing = jest.fn().mockResolvedValue({});
const deletePaymentFromExpenseSharing = jest.fn().mockResolvedValue({});

export {
    authenticate,
    getUserInfo,
    register,
    completeRegistration,
    getAllUserEvents,
    createEvent,
    getEvent,
    saveEvent,
    leaveEvent,
    inviteToEvent,
    bounceFromEvent,
    addExpenseSharingModule,
    addPaymentToExpenseSharing,
    deletePaymentFromExpenseSharing,
    UnexpectedServerError
};
