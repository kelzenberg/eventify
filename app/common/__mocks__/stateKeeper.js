const stateKeeper = jest.requireActual("../stateKeeper");

export const UserContext = stateKeeper.UserContext;
export const isAuthenticated = jest.fn();
export const setCredentials = jest.fn();
export const getAuthenticationToken = jest.fn();
export const getUserInfo = jest.fn();
export const maybeUpdateUserInfo = jest.fn();
export const clearLogin = jest.fn();