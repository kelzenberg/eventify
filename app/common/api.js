import fetcher from "./fetcher";
import * as stateKeeper from "./stateKeeper";

export class UnexpectedServerError extends Error {
    constructor(msg, code) {
        super(msg);
    }
}

export function authenticate(email, password) {
    return new Promise((resolve, reject) => {
        fetcher.request({
            method: "POST",
            path: "authenticate",
            expect: "json",
            body: {"email": email, "password": password},
            allowedStatuses: [fetcher.status.unauthorized]
        })
        .then(([responseData, rawResponse]) => {
            if(rawResponse.status === fetcher.status.unauthorized) {
                reject();
            } else {
                stateKeeper.setCredentials(responseData.token);
                resolve();
            }
        })
        .catch(err => {
            reject(new UnexpectedServerError(err));
        })
    });
}

export function getUserInfo() {
    return fetcher.request({method: "GET", path: "me", expect: "json"}).then(([response]) => response);
}

export function register(email, password, displayName) {
    let body = {
        "email": email,
        "password": password,
        "displayName": displayName
    };
    return fetcher.request({method: "POST", path: "register", expect: "json", body: body})
    .then(([responseData, rawResponse]) => {
        stateKeeper.setCredentials(responseData.token);
        return Promise.resolve();
    })
}

export function logout() {
    console.warn("TODO: api logout");
}

// =========================================== EVENTS =========================================== //

export function getAllUserEvents() {
    return fetcher.request({method: "GET", path: "me/events", expect: "json"}).then(([response]) => response);
}

export function getEvent(eventID) {
    return fetcher.request({method: "GET", path: `events/${eventID}`, expect: "json"}).then(([response]) => response);

    return Promise.resolve({
        "id": "22f7c93e-cfdc-491f-baa2-732fa298d8f6",
        "createdAt": "2021-06-02T15:42:16.478+00:00",
        "updatedAt": "2021-06-02T15:42:16.478+00:00",
        "title": "TestEvent 1",
        "description": "This is a test description",
        "startedAt": "2021-12-24T13:33:37.111+00:00",
        "endedAt": null,
        "expenseSharingModules": [
            {
                "id": "0f994c11-58ba-4718-ae18-5fe099177329",
                "createdAt": "2021-06-02T15:42:40.985+00:00",
                "updatedAt": "2021-06-02T15:42:40.985+00:00",
                "title": "Expense Sharing Module 1",
                "description": "This is a test description",
                "payments": [
                    {
                        "id": "e2ba702a-3713-4804-9b67-de5f9f99b3a8",
                        "createdAt": "2021-06-02T15:43:05.301+00:00",
                        "updatedAt": "2021-06-02T15:43:05.301+00:00",
                        "title": "Payment 1",
                        "amount": 13.37,
                        "payer": {
                            "id": "1798a949-1f72-45ff-b83c-d8e796add4cc",
                            "createdAt": "2021-05-29T17:01:31.716+00:00",
                            "updatedAt": "2021-05-29T17:01:31.716+00:00",
                            "email": "admin@test.de",
                            "displayName": "Admin",
                            "authRole": "ADMIN"
                        },
                        "shareType": "FIXED"
                    }
                ]
            }
        ]
    });
}

export function saveEvent(event) {
    return fetcher.request({method: "PUT", path: `events/${event.id}`, expect: "json"}).then(([response]) => response);
}