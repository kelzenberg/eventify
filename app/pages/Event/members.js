import React from 'react';
import PropTypes from 'prop-types';
import update from 'immutability-helper';
import { ConfirmationDialog, Dialog } from "../../components/Dialog/Dialog";
import { UserContext } from "../../common/stateKeeper";
import * as api from "../../common/api";

export function SmallMembers({event, onEventChanged, setVisibleContent}) {
    const [showDialog, setShowDialog] = React.useState(false);

    return <>
        <div className="card">
            <div className="card-body">
                <div className="d-flex">
                    <p className="fw-bold me-auto">
                        Attendees 
                        <img src="/assets/icons/member-add.svg" className="ps-2" role="button" aria-label="Add new member" data-testid="add member button" onClick={() => setShowDialog(true)}/>
                    </p>
                    <p className="fw-bold text-muted">
                        {event.users.length}
                        <img src="/assets/icons/members.svg" className="ps-2"/>
                    </p>
                </div>
                {event.users.slice(0, 5).map(u => <div className="my-3" key={u.id}>
                    <div className="iconBox iconBox-gray me-2" style={{backgroundImage: "url(/assets/icons/user-image.svg)"}}>
                        {/* <img src={`/api/users/${m.id}/image`}/> */}
                    </div>
                    <span className="align-middle">{u.displayName}</span>
                </div>)}
                <a className="link-primary fw-bold d-block m-2" role="button" onClick={() => setVisibleContent("members")} data-testid="show more button">More...</a>
            </div>
        </div>
        <AddMemberDialog show={showDialog} onHide={() => setShowDialog(false)} eventID={event.id} onEventChanged={onEventChanged}/>
    </>
}

SmallMembers.propTypes = {
    event: PropTypes.object.isRequired,
    onEventChanged: PropTypes.func.isRequired,
    setVisibleContent: PropTypes.func.isRequired
}

export function FullMembers(props) {
    const localUserInfo = React.useContext(UserContext);
    const [addUserVisible, setAddUserVisible] = React.useState(false);
    const [bounceUser, setBounceUser] = React.useState(null);

    function removeMember() {
        api.bounceFromEvent(props.event.id, bounceUser.id)
        .then(() => {
            let newUsers = props.event.users.filter(u => u.id != bounceUser.id);
            setBounceUser(null);
            let newEvent = update(props.event, {users: {$set: newUsers}});
            props.onEventChanged(newEvent);
        })
        .catch(err => {
            console.warn(err);
            props.onErrorMessage("Unfortunately a problem prevented us from saving the event. Please try again.");
            setBounceUser(null);
        })
    }

    return <>
        <a className="link-primary fw-bold d-block m-2" role="button" onClick={() => props.setVisibleContent("modules")} data-testid="close full members button">&lt; Go back to all modules</a>
        <div className="card mb-3">
            <div className="card-body">
                <div className="d-flex">
                    <p className="fw-bold me-auto fs-5">All {props.event.users.length} Attendees: </p>
                    <button type="button" className="btn btn-outline-primary" onClick={() => setAddUserVisible(true)} aria-label="Invite User" data-testid="add member button">Invite...</button>
                </div>
                <div>
                    {props.event.users.map(u => <div className="my-3 d-flex align-items-center" key={u.id}>
                        <div className="iconBox iconBox-gray me-2" style={{backgroundImage: "url(/assets/icons/user-image.svg)"}}>
                            {/* <img src={`/api/users/${m.id}/image`}/> */}
                        </div>
                        <span className="me-auto">{u.displayName}</span>
                        <button hidden={localUserInfo.id == u.id} type="button" role="button" className="btn-close text-end me-2" aria-label="Remove Attendee" onClick={() => setBounceUser(u)} data-testid="remove member button"/>
                    </div>)}
                </div>
            </div>
        </div>
        <ConfirmationDialog 
            show={bounceUser !== null}
            title="Remove this participant?"
            message={`Are you sure you want to remove "${bounceUser === null ? "" : bounceUser.displayName}" from the Event?`}
            confirmText="Remove Participant"
            denyText="Cancel"
            onConfirm={removeMember}
            onDeny={() => setBounceUser(null)}
        />
        <AddMemberDialog show={addUserVisible} onHide={() => setAddUserVisible(false)} eventID={props.event.id} onEventChanged={props.onEventChanged}/>
    </>
}

FullMembers.propTypes = {
    event: PropTypes.object.isRequired,
    onEventChanged: PropTypes.func.isRequired,
    setVisibleContent: PropTypes.func.isRequired,
    onErrorMessage: PropTypes.func.isRequired
}

function AddMemberDialog(props) {
    const [emailAddress, setEmailAddress] = React.useState("");
    const [showValidation, setShowValidation] = React.useState(false);
    const [errorMessage, setErrorMessage] = React.useState(null);

    function reset() {
        props.onHide();
        setEmailAddress("");
        setShowValidation(false);
        setErrorMessage(null);
    }

    function addMember() {
        if(emailAddress === "") {
            setShowValidation(true);
            return;
        }

        api.inviteToEvent(props.eventID, emailAddress)
        .then(eventData => {
            props.onEventChanged(eventData);
            reset();
        })
        .catch(err => {
            console.warn(err);
            setErrorMessage("An unexpected problem prevented us from inviting the person. Please try again.");
        });
    }

    return <Dialog
        show={props.show}
        onHide={reset}
        title="Invite someone to the event"
        closable={true}
        buttons={[{text: "Invite", onClick:addMember}]}
    >
        <form className={showValidation ? "was-validated" : ""}>
            <div className="mb-3">
                <label htmlFor={props.eventID + "_newMemberMail"} className="form-label">E-Mail</label>
                <input type="email" className="form-control" id={props.eventID + "newMemberMail"} required value={emailAddress} onChange={e => setEmailAddress(e.target.value)} data-testid="email input"/>
                <div className="form-text">Enter the E-Mail Address of the person you want to invite.</div>
            </div>
        </form>
        <span className="text-warning fw-bold" hidden={errorMessage === null} data-testid="error message">{errorMessage}</span>
    </Dialog>;
}