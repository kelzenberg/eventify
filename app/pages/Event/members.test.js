import React from 'react';
import {act, fireEvent, getByRole, render, waitFor} from '@testing-library/react';
import "@testing-library/jest-dom/extend-expect";
import { SmallMembers, FullMembers, __RewireAPI__ as rewireMembers } from './members';
import { UserContext } from '../../common/stateKeeper';
import { ConfirmationDialog, Dialog } from '../../components/Dialog/Dialog';
import * as api from "../../common/api";

const AddMemberDialog =  rewireMembers.__get__("AddMemberDialog");

jest.mock("../../components/Dialog/Dialog");
jest.mock("../../common/api");

const localUser = {
    id: "2",
    displayName: "User 2"
};

const event = {
    id: "123456789",
    users: [{
        id: "1",
        displayName: "User 1",
    }, {
        id: "2",
        displayName: "User 2",
    }, {
        id: "3",
        displayName: "User 3",
    }, {
        id: "4",
        displayName: "User 4",
    }, {
        id: "5",
        displayName: "User 5",
    }, {
        id: "6",
        displayName: "User 6",
    }, {
        id: "7",
        displayName: "User 7",
    }]
};

test('Small Members', async () => {
    let dialogMock = jest.fn(() => null);
    rewireMembers.__Rewire__("AddMemberDialog", dialogMock);
    let eventChangeHandler = jest.fn();
    let setVisibleContentHandler = jest.fn();
    let {container, getByTestId} = render(
        <SmallMembers event={event} onEventChanged={eventChangeHandler} setVisibleContent={setVisibleContentHandler}/>
    );

    expect(container).toMatchSnapshot();

    expect(dialogMock).toHaveBeenCalledWith(expect.objectContaining({
        show: false,
        onHide: expect.any(Function),
        eventID: event.id,
        onEventChanged: eventChangeHandler
    }), expect.anything());

    fireEvent.click(getByTestId("show more button"));
    expect(setVisibleContentHandler).toHaveBeenCalledWith("members");

    fireEvent.click(getByTestId("add member button"));
    expect(dialogMock).toHaveBeenCalledWith(expect.objectContaining({
        show: true
    }), expect.anything());

    act(() => getPropsOfLastCall(dialogMock).onHide());

    expect(dialogMock).toHaveBeenLastCalledWith(expect.objectContaining({
        show: false
    }), expect.anything());

    rewireMembers.__ResetDependency__("AddMemberDialog");
});

describe('Full Members', () => {
    let dialogMock = jest.fn(() => null);

    let eventChangeHandler = jest.fn();
    let setVisibleContentHandler = jest.fn();
    let errorMessageHandler = jest.fn();

    function renderComponent() {
        return render(
            <UserContext.Provider value={localUser}>
                <FullMembers event={event} onEventChanged={eventChangeHandler} setVisibleContent={setVisibleContentHandler} onErrorMessage={errorMessageHandler}/>
            </UserContext.Provider>
        );
    }

    beforeAll(() => {
        rewireMembers.__Rewire__("AddMemberDialog", dialogMock);
    });

    afterAll(() => {
        rewireMembers.__ResetDependency__("AddMemberDialog");
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    test('Basic', async () => {
        let {container, getByTestId, getAllByTestId} = renderComponent();
        expect(container).toMatchSnapshot();

        expect(dialogMock).toHaveBeenCalledWith(expect.objectContaining({
            show: false,
            onHide: expect.any(Function),
            eventID: event.id,
            onEventChanged: eventChangeHandler
        }), expect.anything());
    
        expect(ConfirmationDialog).toHaveBeenCalledWith(expect.objectContaining({
            show: false,
            onConfirm: expect.any(Function),
            onDeny: expect.any(Function)
        }), expect.anything());
    });

    test('Add Member Dialog', async () => {
        let {container, getByTestId, getAllByTestId} = renderComponent();
        // show add-member-dialog
        fireEvent.click(getByTestId("add member button"));
        expect(dialogMock).toHaveBeenCalledWith(expect.objectContaining({
            show: true
        }), expect.anything());

        // hide dialog again
        act(() => getPropsOfLastCall(dialogMock).onHide());
        expect(dialogMock).toHaveBeenLastCalledWith(expect.objectContaining({
            show: false
        }), expect.anything());
    });

    test('Remove member but abort', () => {
        let {container, getByTestId, getAllByTestId} = renderComponent();
        let removeButton = getAllByTestId("remove member button")[0];
        fireEvent.click(removeButton);
        expect(ConfirmationDialog).toHaveBeenLastCalledWith(expect.objectContaining({
            show: true
        }), expect.anything());
    
        let dialogProps = getPropsOfLastCall(ConfirmationDialog);
        act(() => dialogProps.onDeny());
        expect(ConfirmationDialog).toHaveBeenLastCalledWith(expect.objectContaining({
            show: false
        }), expect.anything());
    });

    test('Remove member but api fails', async () => {
        let {container, getByTestId, getAllByTestId} = renderComponent();
        let removeButton = getAllByTestId("remove member button")[0];
        let warnSpy = jest.spyOn(console, "warn").mockImplementation(() => {});
        fireEvent.click(removeButton);
        api.bounceFromEvent.mockRejectedValueOnce(new Error("Mock API Error"));
        let dialogProps = getPropsOfLastCall(ConfirmationDialog);
        act(() => dialogProps.onConfirm()); // confirm removal
        expect(api.bounceFromEvent).toHaveBeenLastCalledWith(event.id, event.users[0].id); // check api call

        // dialog should be hidden again
        await waitFor(() => 
            expect(ConfirmationDialog).toHaveBeenLastCalledWith(expect.objectContaining({
                show: false
            }), expect.anything())
        );
        // error should be returned
        expect(errorMessageHandler).toHaveBeenCalled();
        warnSpy.mockRestore();
    });
    
    test('Remove member and succeed', async () => {
        let {container, getByTestId, getAllByTestId} = renderComponent();
        let removeButton = getAllByTestId("remove member button")[0];
        fireEvent.click(removeButton);
        api.bounceFromEvent.mockResolvedValueOnce(null);
        let dialogProps = getPropsOfLastCall(ConfirmationDialog);
        act(() => dialogProps.onConfirm()); // confirm removal
        expect(api.bounceFromEvent).toHaveBeenLastCalledWith(event.id, event.users[0].id); // check api call

        // dialog should be hidden again
        await waitFor(() => 
            expect(ConfirmationDialog).toHaveBeenLastCalledWith(expect.objectContaining({
                show: false
            }), expect.anything())
        );

        expect(errorMessageHandler).not.toHaveBeenCalled();
        expect(eventChangeHandler).toHaveBeenCalled();

        expect(container).toMatchSnapshot();
    });
    
    test('Close', () => {
        let {container, getByTestId, getAllByTestId} = renderComponent();
        fireEvent.click(getByTestId("close full members button"));
        expect(setVisibleContentHandler).toHaveBeenLastCalledWith("modules");
    });
});

describe('Add Member Dialog', () => {
    let hideHandler = jest.fn();
    let eventChangeHandler = jest.fn();

    function renderComponent() {
        return render(
            <AddMemberDialog onHide={hideHandler} onEventChanged={eventChangeHandler} eventID={event.id} show={true}/>
        );
    }

    afterEach(() => {
        jest.clearAllMocks();
    });

    test('Visibility', () => {
        let {rerender, container} = render(
            <AddMemberDialog onHide={hideHandler} onEventChanged={eventChangeHandler} eventID={event.id} show={false}/>
        );

        expect(Dialog).toHaveBeenCalledWith(expect.objectContaining({
            show: false
        }), expect.anything());

        rerender(
            <AddMemberDialog onHide={hideHandler} onEventChanged={eventChangeHandler} eventID={event.id} show={true}/>
        );

        expect(Dialog).toHaveBeenCalledWith(expect.objectContaining({
            show: true
        }), expect.anything());

        expect(container).toMatchSnapshot();
    });

    function prepareWithEmail(email) {
        let {getByTestId} = renderComponent();

        fireEvent.change(getByTestId("email input"), {target: {value: email}});
        let dialogProps = getPropsOfLastCall(Dialog);
        expect(dialogProps).toEqual(expect.objectContaining({
            show: true,
            onHide: expect.any(Function),
            title: expect.any(String),
            closable: true,
            buttons: expect.any(Array)
        }));
        expect(dialogProps.buttons).toHaveLength(1);

        return {addButton: dialogProps.buttons[0].onClick, getByTestId};
    }

    test('Empty email', () => {
        let {addButton} = prepareWithEmail("");
        act(() => addButton());
        expect(api.inviteToEvent).not.toHaveBeenCalled();
    });

    test('API Error', async () => {
        let {addButton, getByTestId} = prepareWithEmail("one@two.three");
        api.inviteToEvent.mockRejectedValueOnce(new Error("Mocked API Error"));
        let warnSpy = jest.spyOn(console, "warn").mockImplementation(() => {});
        act(() => addButton());
        await waitFor(() => expect(getByTestId("error message")).toBeVisible());
        expect(api.inviteToEvent).toHaveBeenCalledWith(event.id, "one@two.three");
        
        expect(eventChangeHandler).not.toHaveBeenCalled();
        warnSpy.mockRestore();
    });

    test('Add New Member', async () => {
        let {addButton, getByTestId} = prepareWithEmail("one@two.three");
        api.inviteToEvent.mockResolvedValueOnce("fake new event data");
        act(() => addButton());
        await waitFor(() => 
            expect(api.inviteToEvent).toHaveBeenCalledWith(event.id, "one@two.three")
        );
        expect(getByTestId("error message")).not.toBeVisible();
        expect(eventChangeHandler).toHaveBeenCalledWith("fake new event data");
    });
});

function getPropsOfLastCall(mockFn){
    return mockFn.mock.calls[mockFn.mock.calls.length-1][0];
}