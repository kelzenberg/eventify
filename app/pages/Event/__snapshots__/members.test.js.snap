// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add Member Dialog Visibility 1`] = `
<div>
  <div>
    Mock Dialog: 
    <form
      class=""
    >
      <div
        class="mb-3"
      >
        <label
          class="form-label"
          for="123456789_newMemberMail"
        >
          E-Mail
        </label>
        <input
          class="form-control"
          data-testid="email input"
          id="123456789newMemberMail"
          required=""
          type="email"
          value=""
        />
        <div
          class="form-text"
        >
          Enter the E-Mail Address of the person you want to invite.
        </div>
      </div>
    </form>
    <span
      class="text-warning fw-bold"
      data-testid="error message"
      hidden=""
    />
  </div>
</div>
`;

exports[`Full Members Basic 1`] = `
<div>
  <a
    class="link-primary fw-bold d-block m-2"
    data-testid="close full members button"
    role="button"
  >
    &lt; Go back to all modules
  </a>
  <div
    class="card mb-3"
  >
    <div
      class="card-body"
    >
      <div
        class="d-flex"
      >
        <p
          class="fw-bold me-auto fs-5"
        >
          All 
          7
           Attendees: 
        </p>
        <button
          aria-label="Invite User"
          class="btn btn-outline-primary"
          data-testid="add member button"
          type="button"
        >
          Invite...
        </button>
      </div>
      <div>
        <div
          class="my-3 d-flex align-items-center"
        >
          <div
            class="iconBox iconBox-gray me-2"
            style="background-image: url(/assets/icons/user-image.svg);"
          />
          <span
            class="me-auto"
          >
            User 1
          </span>
          <button
            aria-label="Remove Attendee"
            class="btn-close text-end me-2"
            data-testid="remove member button"
            role="button"
            type="button"
          />
        </div>
        <div
          class="my-3 d-flex align-items-center"
        >
          <div
            class="iconBox iconBox-gray me-2"
            style="background-image: url(/assets/icons/user-image.svg);"
          />
          <span
            class="me-auto"
          >
            User 2
          </span>
          <button
            aria-label="Remove Attendee"
            class="btn-close text-end me-2"
            data-testid="remove member button"
            hidden=""
            role="button"
            type="button"
          />
        </div>
        <div
          class="my-3 d-flex align-items-center"
        >
          <div
            class="iconBox iconBox-gray me-2"
            style="background-image: url(/assets/icons/user-image.svg);"
          />
          <span
            class="me-auto"
          >
            User 3
          </span>
          <button
            aria-label="Remove Attendee"
            class="btn-close text-end me-2"
            data-testid="remove member button"
            role="button"
            type="button"
          />
        </div>
        <div
          class="my-3 d-flex align-items-center"
        >
          <div
            class="iconBox iconBox-gray me-2"
            style="background-image: url(/assets/icons/user-image.svg);"
          />
          <span
            class="me-auto"
          >
            User 4
          </span>
          <button
            aria-label="Remove Attendee"
            class="btn-close text-end me-2"
            data-testid="remove member button"
            role="button"
            type="button"
          />
        </div>
        <div
          class="my-3 d-flex align-items-center"
        >
          <div
            class="iconBox iconBox-gray me-2"
            style="background-image: url(/assets/icons/user-image.svg);"
          />
          <span
            class="me-auto"
          >
            User 5
          </span>
          <button
            aria-label="Remove Attendee"
            class="btn-close text-end me-2"
            data-testid="remove member button"
            role="button"
            type="button"
          />
        </div>
        <div
          class="my-3 d-flex align-items-center"
        >
          <div
            class="iconBox iconBox-gray me-2"
            style="background-image: url(/assets/icons/user-image.svg);"
          />
          <span
            class="me-auto"
          >
            User 6
          </span>
          <button
            aria-label="Remove Attendee"
            class="btn-close text-end me-2"
            data-testid="remove member button"
            role="button"
            type="button"
          />
        </div>
        <div
          class="my-3 d-flex align-items-center"
        >
          <div
            class="iconBox iconBox-gray me-2"
            style="background-image: url(/assets/icons/user-image.svg);"
          />
          <span
            class="me-auto"
          >
            User 7
          </span>
          <button
            aria-label="Remove Attendee"
            class="btn-close text-end me-2"
            data-testid="remove member button"
            role="button"
            type="button"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Full Members Remove member and succeed 1`] = `
<div>
  <a
    class="link-primary fw-bold d-block m-2"
    data-testid="close full members button"
    role="button"
  >
    &lt; Go back to all modules
  </a>
  <div
    class="card mb-3"
  >
    <div
      class="card-body"
    >
      <div
        class="d-flex"
      >
        <p
          class="fw-bold me-auto fs-5"
        >
          All 
          7
           Attendees: 
        </p>
        <button
          aria-label="Invite User"
          class="btn btn-outline-primary"
          data-testid="add member button"
          type="button"
        >
          Invite...
        </button>
      </div>
      <div>
        <div
          class="my-3 d-flex align-items-center"
        >
          <div
            class="iconBox iconBox-gray me-2"
            style="background-image: url(/assets/icons/user-image.svg);"
          />
          <span
            class="me-auto"
          >
            User 1
          </span>
          <button
            aria-label="Remove Attendee"
            class="btn-close text-end me-2"
            data-testid="remove member button"
            role="button"
            type="button"
          />
        </div>
        <div
          class="my-3 d-flex align-items-center"
        >
          <div
            class="iconBox iconBox-gray me-2"
            style="background-image: url(/assets/icons/user-image.svg);"
          />
          <span
            class="me-auto"
          >
            User 2
          </span>
          <button
            aria-label="Remove Attendee"
            class="btn-close text-end me-2"
            data-testid="remove member button"
            hidden=""
            role="button"
            type="button"
          />
        </div>
        <div
          class="my-3 d-flex align-items-center"
        >
          <div
            class="iconBox iconBox-gray me-2"
            style="background-image: url(/assets/icons/user-image.svg);"
          />
          <span
            class="me-auto"
          >
            User 3
          </span>
          <button
            aria-label="Remove Attendee"
            class="btn-close text-end me-2"
            data-testid="remove member button"
            role="button"
            type="button"
          />
        </div>
        <div
          class="my-3 d-flex align-items-center"
        >
          <div
            class="iconBox iconBox-gray me-2"
            style="background-image: url(/assets/icons/user-image.svg);"
          />
          <span
            class="me-auto"
          >
            User 4
          </span>
          <button
            aria-label="Remove Attendee"
            class="btn-close text-end me-2"
            data-testid="remove member button"
            role="button"
            type="button"
          />
        </div>
        <div
          class="my-3 d-flex align-items-center"
        >
          <div
            class="iconBox iconBox-gray me-2"
            style="background-image: url(/assets/icons/user-image.svg);"
          />
          <span
            class="me-auto"
          >
            User 5
          </span>
          <button
            aria-label="Remove Attendee"
            class="btn-close text-end me-2"
            data-testid="remove member button"
            role="button"
            type="button"
          />
        </div>
        <div
          class="my-3 d-flex align-items-center"
        >
          <div
            class="iconBox iconBox-gray me-2"
            style="background-image: url(/assets/icons/user-image.svg);"
          />
          <span
            class="me-auto"
          >
            User 6
          </span>
          <button
            aria-label="Remove Attendee"
            class="btn-close text-end me-2"
            data-testid="remove member button"
            role="button"
            type="button"
          />
        </div>
        <div
          class="my-3 d-flex align-items-center"
        >
          <div
            class="iconBox iconBox-gray me-2"
            style="background-image: url(/assets/icons/user-image.svg);"
          />
          <span
            class="me-auto"
          >
            User 7
          </span>
          <button
            aria-label="Remove Attendee"
            class="btn-close text-end me-2"
            data-testid="remove member button"
            role="button"
            type="button"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Small Members 1`] = `
<div>
  <div
    class="card"
  >
    <div
      class="card-body"
    >
      <div
        class="d-flex"
      >
        <p
          class="fw-bold me-auto"
        >
          Attendees
          <img
            aria-label="Add new member"
            class="ps-2"
            data-testid="add member button"
            role="button"
            src="/assets/icons/member-add.svg"
          />
        </p>
        <p
          class="fw-bold text-muted"
        >
          7
          <img
            class="ps-2"
            src="/assets/icons/members.svg"
          />
        </p>
      </div>
      <div
        class="my-3"
      >
        <div
          class="iconBox iconBox-gray me-2"
          style="background-image: url(/assets/icons/user-image.svg);"
        />
        <span
          class="align-middle"
        >
          User 1
        </span>
      </div>
      <div
        class="my-3"
      >
        <div
          class="iconBox iconBox-gray me-2"
          style="background-image: url(/assets/icons/user-image.svg);"
        />
        <span
          class="align-middle"
        >
          User 2
        </span>
      </div>
      <div
        class="my-3"
      >
        <div
          class="iconBox iconBox-gray me-2"
          style="background-image: url(/assets/icons/user-image.svg);"
        />
        <span
          class="align-middle"
        >
          User 3
        </span>
      </div>
      <div
        class="my-3"
      >
        <div
          class="iconBox iconBox-gray me-2"
          style="background-image: url(/assets/icons/user-image.svg);"
        />
        <span
          class="align-middle"
        >
          User 4
        </span>
      </div>
      <div
        class="my-3"
      >
        <div
          class="iconBox iconBox-gray me-2"
          style="background-image: url(/assets/icons/user-image.svg);"
        />
        <span
          class="align-middle"
        >
          User 5
        </span>
      </div>
      <a
        class="link-primary fw-bold d-block m-2"
        data-testid="show more button"
        role="button"
      >
        More...
      </a>
    </div>
  </div>
</div>
`;
